### PR TITLE
[3/3][AOT][DeviceAPI] Wire up cpacked Device API context

### PR DESCRIPTION
**Pull request by Mousius**
Adding the same functionality for the Device API to the cpacked calling convention. The MakePackedAPI pass now implicitly uses any variable named `kDeviceContextVar` as the `resource_handle` and this is then used in the `cpacked` calling convention which always expects some form of resource_handle to be passed.

This is stacked on top of #9500 and #9395

### CODE REVIEW

File: src/relay/backend/aot_executor_codegen.cc [hunk 1]
```diff
@@ -347,18 +347,30 @@ class AOTExecutorCodegen : public MixedModeVisitor {
     }
 
     GlobalVar global_var = call_lowered_props.lowered_func;
+    tir::Var empty_var("no_device_context", DataType::Handle());
     bool has_c_device_api_context = device_contexts_.count(global_var) != 0;
+    bool use_cpacked_api = !use_unpacked_api_;
+
     if (has_c_device_api_context) {
```
Comment (mbs-octoml) [inline src/relay/backend/aot_executor_codegen.cc:354]: Let's comment these three semi-implicit calling conventions here, including a one-sentence summary of where the extra args show up in the runtime.

**Review by mbs-octoml — APPROVED** [review]

**Review by manupa-arm — COMMENTED** [review]
Broadly looks good!. Just one question but could do in a follow up since this is green.

File: src/relay/backend/aot_executor_codegen.cc [hunk 2]
```diff
@@ -347,18 +347,41 @@ class AOTExecutorCodegen : public MixedModeVisitor {
...
           GenerateDeviceHook(context, "Open"),
           func_call,
           GenerateDeviceHook(context, "Close"),
       }));
+    } else if (use_cpacked_api) {
+      // call_cpacked calling convention needs a blank context
+      args.push_back(empty_var);
+
+      tir::Evaluate func_call(tvm::tir::Call(DataType::Int(32), calling_pattern, args));
+      tir::LetStmt set_zero(empty_var, tir::make_zero(DataType::Handle()), func_call);
```
[the remark this answers is not in the document]
Comment (Mousius) [reply src/relay/backend/aot_executor_codegen.cc]: Good point @manupa-arm, I've removed it :smile_cat:

**Review by manupa-arm — APPROVED** [review]
Thanks @Mousius . LGTM

**Comment by manupa-arm** [comment]
Thanks @mbs-octoml @Mousius . This is merged now!

**Pull request merged**